tokio-macros/src/entry.rs

```diff
@@ -1,11 +1,171 @@
+            _ => Err(format!("No such runtime flavor `{}`.", s)),
+        }
+    }
+}
+
+struct FinalConfig {
+    flavor: RuntimeFlavor,
+    num_workers: Option<usize>,
+    max_threads: Option<usize>,
+}
+
+struct Configuration {
+    rt_threaded_available: bool,
+    default_flavor: RuntimeFlavor,
+    flavor: Option<RuntimeFlavor>,
+    num_workers: Option<(usize, Span)>,
+    max_threads: Option<(usize, Span)>,
+}
+impl Configuration {
+    fn new(is_test: bool, rt_threaded: bool) -> Self {
+        Configuration {
+            rt_threaded_available: rt_threaded,
+            default_flavor: match is_test {
+                true => RuntimeFlavor::InPlace,
+                false => RuntimeFlavor::WorkStealing,
+            },
+            flavor: None,
+            num_workers: None,
+            max_threads: None,
+        }
+    }
+    fn set_flavor(
+        &mut self,
+        runtime: syn::Lit,
+        span: Span,
+    ) -> Result<(), syn::Error> {
+        if self.flavor.is_some() {
+            return Err(syn::Error::new(span, "`flavor` set multiple times."));
+        }
+
+        let runtime_str = parse_string(runtime, span, "flavor")?;
+        let runtime = RuntimeFlavor::from_str(&runtime_str)
+            .map_err(|err| syn::Error::new(span, err))?;
+        self.flavor = Some(runtime);
+        Ok(())
+    }
+    fn set_num_workers(
+        &mut self,
+        num_workers: syn::Lit,
+        span: Span,
+    ) -> Result<(), syn::Error> {
+        if self.num_workers.is_some() {
+            return Err(syn::Error::new(span, "`num_workers` set multiple times."));
+        }
+
+        let num_workers = parse_int(num_workers, span, "num_workers")?;
+        if num_workers == 0 {
+            return Err(syn::Error::new(span, "`num_workers` may not be 0."));
+        }
+        self.num_workers = Some((num_workers, span));
+        Ok(())
+    }
+    fn set_max_threads(
+        &mut self,
+        max_threads: syn::Lit,
+        span: Span,
+    ) -> Result<(), syn::Error> {
+        if self.max_threads.is_some() {
+            return Err(syn::Error::new(span, "`max_threads` set multiple times."));
+        }
+
+        let max_threads = parse_int(max_threads, span, "max_threads")?;
+        if max_threads == 0 {
+            return Err(syn::Error::new(span, "`max_threads` may not be 0."));
+        }
+        self.max_threads = Some((max_threads, span));
+        Ok(())
+    }
+    fn build(&self) -> Result<FinalConfig, syn::Error> {
+        let flavor = self.flavor.unwrap_or(self.default_flavor);
+        match flavor {
+            RuntimeFlavor::InPlace => {
+                if let Some((_, num_workers_span)) = &self.num_workers {
+                    return Err(syn::Error::new(
+                        num_workers_span.clone(),
+                        "The `num_workers` option requires the `work_stealing` runtime flavor."
+                    ));
+                }
+                assert!(self.num_workers.is_none());
```

Comment:
can we statically ensure this somehow? Maybe via enum values?

Reply:
We could certainly statically ensure it in the `FinalConfig`, but that wouldn't remove this assert. I suppose we could turn `Configuration` into a state machine of some sort.

Reply:
Though turning `Configuration` into a state machine is awkward. I chose the current design because it allows e.g. setting `num_workers` first, then setting the `flavor`.

Reply:
Here's one option:
```rust
match (flavor, self.num_workers) {
    (CurrentThread, Some((_, num_workers_span))) => {
        Err(syn::Error::new(
            num_workers_span.clone(),
            "The `num_workers` option requires the `threaded` runtime flavor."
        ))
    },
    (CurrentThread, None) => {
        Ok(FinalConfig {
            flavor,
            num_workers: None,
        })
    },
    (Threaded, num_workers) if self.rt_threaded_available => {
        Ok(FinalConfig {
            flavor,
            num_workers: num_workers.map(|(val, _span)| val),
        })
    },
    (Threaded, _) => {
        let msg = if self.flavor.is_none() {
            "The default runtime flavor is `threaded`, but the `rt-threaded` feature is disabled."
        } else {
            "The runtime flavor `threaded` requires the `rt-threaded` feature."
        };
        Err(syn::Error::new(Span::call_site(), msg))
    },
}
```